### PR TITLE
[Fault tolerance] Make executor actor restartable and fetch data from blockManager in actor calls

### PR DESCRIPTION
**Pull request by kira-lin**

### CODE REVIEW

**Comment by kira-lin** [comment]
If we use a normal task, which calls the actor internally, the actor method still need to be resubmitted because it's a dependency. I'm not sure if using spark to call the executors will work, but I think sparkcontext would be needed to do so.

**Review by carsonwang — COMMENTED** [review]
Thanks for the work!

File: core/src/main/scala/org/apache/spark/deploy/raydp/ApplicationInfo.scala [hunk 1]
```diff
@@ -138,6 +142,12 @@ private[spark] class ApplicationInfo(
     registeredExecutors
   }
 
+  def isRemovedExecutor(executorId: String): Boolean = {
```
[the remark this answers is not in the document]
Comment (kira-lin) [reply core/src/main/scala/org/apache/spark/deploy/raydp/ApplicationInfo.scala:145]: This was used in previous implementation. I'll remove it

File: core/src/main/scala/org/apache/spark/deploy/raydp/RayAppMaster.scala [hunk 2]
```diff
@@ -177,6 +183,25 @@ class RayAppMaster(host: String,
...
         }
         context.reply(success)
+
+      case RequestAddPendingRestartedExecutor(executorId) =>
+        if (appInfo.remainingUnRegisteredExecutors > 0) {
+          val cores = appInfo.desc.coresPerExecutor.getOrElse(1)
+          val memory = appInfo.desc.memoryPerExecutorMB
+          val newExecutorId = s"${appInfo.getNextExecutorId()}"
+          // ray actor will restart using the old ID
+          val handlerOpt = Ray.getActor("raydp-executor-" + executorId)
```
[the remark this answers is not in the document]
Comment (kira-lin) [reply core/src/main/scala/org/apache/spark/deploy/raydp/RayAppMaster.scala:193]: No matter how many time it has failed, it is restarted using the parameters given when first created. The name won't change, either.

File: python/raydp/spark/ray_cluster_master.py [hunk 3]
```diff
@@ -27,7 +28,11 @@
 from copy import copy
 import glob
 import ray
-from py4j.java_gateway import JavaGateway, GatewayParameters
+from py4j.java_gateway import java_import, JavaGateway, GatewayParameters
+from py4j.clientserver import ClientServer, JavaParameters, PythonParameters
+from pyspark.find_spark_home import _find_spark_home
```
[the remark this answers is not in the document]
Comment (kira-lin) [reply python/raydp/spark/ray_cluster_master.py:33]: ok

File: core/src/main/scala/org/apache/spark/executor/RayCoarseGrainedExecutorBackend.scala [hunk 4]
```diff
@@ -259,4 +285,90 @@ class RayCoarseGrainedExecutorBackend(
...
+      }
+    }
+    val success = driverAgent.askSync[Boolean](RecacheRDD(rddId))
+    env.shutdown
+  }
+
+  def getRDDPartition(rddId: Int,
+                      partitionId: Int,
+                      schemaStr: String,
+                      driverAgentUrl: String): Array[Byte] = {
```
[the remark this answers is not in the document]
Comment (kira-lin) [reply core/src/main/scala/org/apache/spark/executor/RayCoarseGrainedExecutorBackend.scala:333]: ok

File: core/src/main/scala/org/apache/spark/deploy/raydp/RayAppMaster.scala [hunk 5]
```diff
@@ -177,6 +183,25 @@ class RayAppMaster(host: String,
           }
         }
         context.reply(success)
+
+      case RequestAddPendingRestartedExecutor(executorId) =>
+        if (appInfo.remainingUnRegisteredExecutors > 0) {
+          val cores = appInfo.desc.coresPerExecutor.getOrElse(1)
+          val memory = appInfo.desc.memoryPerExecutorMB
+          val newExecutorId = s"${appInfo.getNextExecutorId()}"
```
[the remark this answers is not in the document]
Comment (kira-lin) [reply core/src/main/scala/org/apache/spark/deploy/raydp/RayAppMaster.scala:191]: right

File: core/src/main/scala/org/apache/spark/deploy/raydp/RayAppMaster.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (carsonwang) [reply core/src/main/scala/org/apache/spark/deploy/raydp/RayAppMaster.scala:193]: After the first restart, I saw the executorId was modified in RayCoarseGrainedExecutorBackend? Where is the original Id stored and can be got?

[the remark this answers is not in the document]
Comment (kira-lin) [reply core/src/main/scala/org/apache/spark/deploy/raydp/RayAppMaster.scala:193]: Yes, it is changed. The original id is not stored in executor. It is saved in a mapping in RayAppMaster restartedExecutors. Ray also stores it in task's lineage.

**Pull request merged**